crates/ethereum/src/lib.rs

```diff
@@ -1,111 +1,434 @@
-pub mod log;
-pub mod provider;
-pub mod state_update;
+#[derive(Clone)]
+pub struct EthereumClient {
+    url: Url,
+    http: Client,
+}
 
-#[derive(Debug, Clone, PartialEq, Hash, Eq)]
-pub struct BlockOrigin {
-    pub hash: EthereumBlockHash,
-    pub number: EthereumBlockNumber,
+impl EthereumClient {
+    pub fn new(url: reqwest::Url) -> Self {
+        let http_client = reqwest::ClientBuilder::new()
+            .build()
+            .expect("reqwest HTTP client");
+        Self {
+            url,
+            http: http_client,
+        }
+    }
+
+    pub fn new_with_password(mut url: reqwest::Url, password: &str) -> anyhow::Result<Self> {
+        url.set_password(Some(password))
+            .map_err(|_| anyhow::anyhow!("Setting password failed"))?;
+        let http_client = reqwest::ClientBuilder::new()
+            .build()
+            .expect("reqwest HTTP client");
+        Ok(Self {
+            url,
+            http: http_client,
+        })
+    }
 }
 
-#[derive(Debug, Clone, PartialEq, Hash, Eq)]
-pub struct TransactionOrigin {
-    pub hash: EthereumTransactionHash,
-    pub index: EthereumTransactionIndex,
+impl StarknetEthereumClient {
+    pub fn new(client: EthereumClient, l1_addr: EthereumAddress) -> Self {
+        Self {
+            l1_addr,
+            eth: client,
+        }
+    }
 }
 
-/// An Ethereum origin point.
-#[derive(Debug, Clone, PartialEq, Hash, Eq)]
-pub struct EthOrigin {
-    pub block: BlockOrigin,
-    pub transaction: TransactionOrigin,
-    pub log_index: EthereumLogIndex,
+pub mod core_contract {
+    use const_decoder::Decoder;
+
+    // Sources of L1 contract addresses:
+    //
+    // $ curl https://alpha-mainnet.starknet.io/feeder_gateway/get_contract_addresses
+    // {"GpsStatementVerifier": "0x47312450B3Ac8b5b8e247a6bB6d523e7605bDb60", "Starknet": "0xc662c410C0ECf747543f5bA90660f6ABeBD9C8c4"}
+    //
+    // $ curl https://alpha4.starknet.io/feeder_gateway/get_contract_addresses
+    // {"GpsStatementVerifier": "0x8f97970aC5a9aa8D130d35146F5b59c4aef57963", "Starknet": "0xde29d060D45901Fb19ED6C6e959EB22d8626708e"}
+    //
+    // $ curl https://alpha4-2.starknet.io/feeder_gateway/get_contract_addresses
+    // {"GpsStatementVerifier": "0x8f97970aC5a9aa8D130d35146F5b59c4aef57963", "Starknet": "0xa4eD3aD27c294565cB0DCc993BDdCC75432D498c"}
+    //
+    // $ curl https://external.integration.starknet.io/feeder_gateway/get_contract_addresses
+    // {"GpsStatementVerifier": "0x8f97970aC5a9aa8D130d35146F5b59c4aef57963", "Starknet": "0xd5c325D183C592C94998000C5e0EED9e6655c020"}
+
+    pub const MAINNET: [u8; 20] = Decoder::Hex.decode(b"c662c410C0ECf747543f5bA90660f6ABeBD9C8c4");
+    pub const TESTNET: [u8; 20] = Decoder::Hex.decode(b"de29d060D45901Fb19ED6C6e959EB22d8626708e");
+    pub const TESTNET2: [u8; 20] = Decoder::Hex.decode(b"a4eD3aD27c294565cB0DCc993BDdCC75432D498c");
+    pub const INTEGRATION: [u8; 20] =
+        Decoder::Hex.decode(b"d5c325D183C592C94998000C5e0EED9e6655c020");
```

Comment:
Is it worth another dependency for this?

Is it not enough to just use `[0xc6, 0x62, 0xc4...]`

Reply:
It is a tiny dependency that allows `const decode`. IMHO hex string is way more readable (and searchable) that raw byte slice.

Reply:
Ah I see all the deps are dev deps. Okay 👍 Could be worth using in our `felt!` macro instead of the hex parsing we do instead.